src/JitInterface/src/CorInfoImpl.cs

```diff
@@ -2023,37 +2142,83 @@ private void embedGenericHandle(ref CORINFO_RESOLVED_TOKEN pResolvedToken, [Mars
+            {
+                pResult.lookup.lookupKind.needsRuntimeLookup = true;
+                pResult.lookup.runtimeLookup.signature = null;
+                pResult.lookup.runtimeLookup.indirections = CORINFO.USEHELPER;
 
-                // If the target is not shared then we've already got our result and
-                // can simply do a static look up
-                pResult.lookup.lookupKind.needsRuntimeLookup = false;
+                MethodDesc contextMethod = methodFromContext(pResolvedToken.tokenContext);
 
-                if (pResolvedToken.tokenType == CorInfoTokenKind.CORINFO_TOKENKIND_NewObj)
+                // Do not bother computing the runtime lookup if we are inlining. The JIT is going
+                // to abort the inlining attempt anyway.
+                if (contextMethod != MethodBeingCompiled)
+                    return;
+
+                if (contextMethod.RequiresInstMethodDescArg())
                 {
-                    pResult.lookup.constLookup.handle = (CORINFO_GENERIC_STRUCT_*)ObjectToHandle(_compilation.NodeFactory.ConstructedTypeSymbol(td));
+                    pResult.lookup.lookupKind.runtimeLookupKind = CORINFO_RUNTIME_LOOKUP_KIND.CORINFO_LOOKUP_METHODPARAM;
                 }
                 else
-                    pResult.lookup.constLookup.handle = (CORINFO_GENERIC_STRUCT_*)ObjectToHandle(_compilation.NodeFactory.NecessaryTypeSymbol(td));
-                pResult.lookup.constLookup.accessType = InfoAccessType.IAT_VALUE;
+                {
+                    if (contextMethod.RequiresInstMethodTableArg())
+                        pResult.lookup.lookupKind.runtimeLookupKind = CORINFO_RUNTIME_LOOKUP_KIND.CORINFO_LOOKUP_CLASSPARAM;
+                    else
+                        pResult.lookup.lookupKind.runtimeLookupKind = CORINFO_RUNTIME_LOOKUP_KIND.CORINFO_LOOKUP_THISOBJ;
+                }
             }
+            else
+            {
+                // If the target is not shared then we've already got our result and
+                // can simply do a static look up
+                pResult.lookup.lookupKind.needsRuntimeLookup = false;
 
-            Debug.Assert(pResult.compileTimeHandle != null);
+                pResult.lookup.constLookup.accessType = InfoAccessType.IAT_VALUE;
+            }
         }
 
         private void getLocationOfThisType(out CORINFO_LOOKUP_KIND result, CORINFO_METHOD_STRUCT_* context)
         {
             result = new CORINFO_LOOKUP_KIND();
-            result.needsRuntimeLookup = false;
-            result.runtimeLookupKind = CORINFO_RUNTIME_LOOKUP_KIND.CORINFO_LOOKUP_THISOBJ;
 
-            // TODO: shared generics
+            MethodDesc method = HandleToObject(context);
+
+            if (method.IsSharedByGenericInstantiations)
+            {
+                result.needsRuntimeLookup = true;
+
+                // If we've got a vtable extra argument, go through that 
+                if (method.RequiresInstMethodTableArg())
```

Comment:
It may be nice to give it same structure as above (or move it to a helper function).